tritium-jmh/src/jmh/java/com/palantir/tritium/microbenchmarks/MetricsLifecycleBenchmark.java

```diff
@@ -0,0 +1,110 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.tritium.microbenchmarks;
+
+import com.palantir.tritium.metrics.registry.DefaultTaggedMetricRegistry;
+import com.palantir.tritium.metrics.registry.MetricName;
+import com.palantir.tritium.metrics.registry.TaggedMetricRegistry;
+import java.util.concurrent.TimeUnit;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.Warmup;
+import org.openjdk.jmh.profile.GCProfiler;
+import org.openjdk.jmh.profile.JavaFlightRecorderProfiler;
+import org.openjdk.jmh.runner.Runner;
+import org.openjdk.jmh.runner.options.OptionsBuilder;
+
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(TimeUnit.MICROSECONDS)
+@Warmup(iterations = 3, time = 1, timeUnit = TimeUnit.SECONDS)
+@Measurement(iterations = 3, time = 1, timeUnit = TimeUnit.SECONDS)
+@Fork(1)
+@State(Scope.Benchmark)
+@SuppressWarnings({"designforextension", "NullAway", "PreferJavaTimeOverload"})
+public class MetricsLifecycleBenchmark {
+    private static final MetricName counterName =
+            MetricName.builder().safeName("counter").putSafeTags("tag", "value").build();
+    private static final MetricName histogramName = MetricName.builder()
+            .safeName("histogram")
+            .putSafeTags("tag", "value")
+            .build();
+    private static final MetricName meterName =
+            MetricName.builder().safeName("meter").putSafeTags("tag", "value").build();
+    private static final MetricName timerName =
+            MetricName.builder().safeName("timer").putSafeTags("tag", "value").build();
+    private TaggedMetricRegistry metrics;
+
+    @Setup
+    public void before() {
+        metrics = constructBaseRegistry();
+    }
+
+    @Benchmark
+    public void counter() {
+        metrics.counter(counterName).inc();
+    }
+
+    @Benchmark
+    public void histogram() {
+        metrics.histogram(histogramName).update(1);
+    }
+
+    @Benchmark
+    public void meter() {
+        metrics.meter(meterName).mark();
+    }
+
+    @Benchmark
+    public void timer() {
+        metrics.timer(counterName).update(1_000_000L, TimeUnit.NANOSECONDS);
+    }
+
+    private static TaggedMetricRegistry constructBaseRegistry() {
+        TaggedMetricRegistry registry = new DefaultTaggedMetricRegistry();
+        for (int i = 0; i < 50; i++) {
+            registry.addMetrics("registry id", Integer.toString(i), constructSubRegistry());
+        }
```

Comment:
I don't think any of the subregistries are checked when we call `AbstractTaggedMetricRegistry.getOrAdd, only the main concurrenthashmap -- we may want to add metrics directly to the top-level registry

Reply:
moving to single registry